README.md
title: "Waste over Time"
author: "V.M. van Straalen"
date: "August 23, 2017"
output: html_document
---

![Statistics Netherlands](https://github.com/Statistics-Netherlands/ewaste/blob/master/images/cbs_logo_en.png)

# Table of Contents
<!-- Table of contents generated generated by http://tableofcontent.eu -->
- [Purpose](#purpose)
- [Installation](#installation)
- [Usage](#usage)
- [Source file listing](#source-file-listing)
- [Output](#output)
- [Description of the program components](#description-of-the-program-components)
  - [Preparing source data sets of prodcom and international trade](#preparing-source-data-sets-of-prodcom-and-international-trade)
  - [Estimation of confidential prodcom data.](#estimation-of-confidential-prodcom-data)
      - [Step 1 - Ratio between units exported and units produced over years](#step-1-ratio-between-units-exported-and-units-produced-over-years)
      - [Step 2 - Ratio between units exported and units produced over countries](#step-2-ratio-between-units-exported-and-units-produced-over-countries)
      - [Finalize - Fitting the results](#finalize-fitting-the-results)
  - [Calculate apparent consumption](#calculate-apparent-consumption)
  - [Description of method to calculate POM (Put on market)](#description-of-method-to-calculate-pom-put-on-market)
    - [First adjustments](#first-adjustments)
    - [Manual corrections](#manual-corrections)
    - [Solar panel data](#solar-panel-data)
    - [Automatic corrections for all UNU_Keys, countries and years](#automatic-corrections-for-all-unu_keys-countries-and-years)
      - [Remove unreliable data by comparing with other years from the same Country and product category](#remove-unreliable-data-by-comparing-with-other-years-from-the-same-country-and-product-category)
      - [Remove unreliable data by comparing data on a specific year and product category with other countries](#remove-unreliable-data-by-comparing-data-on-a-specific-year-and-product-category-with-other-countries)
      - [Calculate years by using the average of other years.](#calculate-years-by-using-the-average-of-other-years)
      - [Calculate missing years by using the average of other similar countries](#calculate-missing-years-by-using-the-average-of-other-similar-countries)
      - [Eliminate big changes between years](#eliminate-big-changes-between-years)
    - [Correction of CRT monitors (UNU_Key 0302)](#correction-of-crt-monitors-unu_key-0302)
    - [Calculate past POM](#calculate-past-pom)
    - [Calculate future POM](#calculate-future-pom)
    - [Result](#result)
  - [Calculation of WEEE generated](#calculation-of-weee-generated)
  - [Calculated aggregates, stock, graphs and output files](#calculated-aggregates-stock-graphs-and-output-files)
- [Reference](#reference)
- [License](#license)
- [Annex 1 - Instruction to update International Trade source datafile](#annex-1-instruction-to-update-international-trade-source-datafile)
  - [the Eurostat Bulk Download facility method](#the-eurostat-bulk-download-facility-method)
  - [Comext website method](#comext-website-method)
- [Annex 2 - Extrapolation methods](#annex-2-extrapolation-methods)
  - [KPI_PPP](#kpi_ppp)
  - [ConnectionConstant](#connectionconstant)
  - [KPI](#kpi)
  - [LastReliableConstant](#lastreliableconstant)
  - [Extrapolation Exceptions](#extrapolation-exceptions)


<a name="purpose"/>
# Purpose
The Waste over Time (or WOT) script is a script developed by Statistics Netherlands. It uses European production and trade statistics with the 'apparent consumption method' to estimate sales in weight of products. The script also comes with data on life times of electronic and electric equipment, and allows to calculate the urban mine and to forecast waste generated. 

Results from the calculations can be viewed [here](https://statistics-netherlands.shinyapps.io/sales_and_waste/): <BR>
[![screenshot](https://github.com/Statistics-Netherlands/ewaste/blob/master/images/screenshot.png)](https://statistics-netherlands.shinyapps.io/sales_and_waste/)


The script here available at the GitHub shows the calculations to produce this output.

The script can be used for: 

**WEEE-Directive:**

* The methodology in the script follows the so-called 'common methodology' to determine Placed on the Market and WEEE Generated as defined in [article 7 of the EU-WEEE Directive](http://eur-lex.europa.eu/legal-content/NL/TXT/?uri=uriserv:OJ.L_.2012.197.01.0038.01.NLD).
For more information regarding the WEEE-Directive: [WEEE-Directive](http://ec.europa.eu/environment/waste/weee/index_en.htm)
* Governmental authorities of the EU Member States, inspection agencies and WEEE-registers can use the script to obtain an alternative to equipment placed on the market, and forecast waste streams.


**Circular Economy**

* The script has been used in the [H2020 ProSUM Project](http://www.prosumproject.eu/).
* The script allows to study Material Flows and Urban Mining by relating sales, use phase, and forecasts of waste streams. It can be easily extended to other materials. 


The WOT script contains all relevant source data to perform the calculations for each EU-28 Member State for 54 types of electronical and electrical equipment.  The statistical source data comes from Eurostat. The other parameters come from a results of a project of DG Environment in which Statistics Netherlands participated. [Art7_publication](http://ec.europa.eu/environment/waste/weee/pdf/Final_Report_Art7_publication.pdf). 

The script has been designed that it can be easily extended to other types of products. The script has been written in [R](https://cran.r-project.org).

<br>
<br>
<br>

<a name="installation"/>
# Installation
**R**

No R knowledge is necessary to run this series of scripts. But R needs to be installed on your computer. For more information on how to install R see [R Installation and Administration](https://cran.r-project.org/doc/manuals/r-release/R-admin.html)

The scripts work with R version 3.2.3. Only script [00a_Prepare_Prodcom_data.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/00a_Prepare_Prodcom_data.R), which is not part of the main program, needs at least R version 3.3.3.

In case you wish to see and understand all the calculations, knowledge of the R language is of course needed.

<br>

**Wast over Time (WOT) script**

From the GitHub e-waste repository (you are there now when reading this file) click on the menu item 'releases' in the middle of the screen. Click on 'zip' behind the most recent release. Save this file on your computer and unpack. Go to the folder .\ewaste-master\scripts. Copy that location to the clipboard. 

Now open de command line (Windows: cmd.exe, aka 'DOS box') by typing cmd into the windows search box and pressing enter.
Here type the command SETX EWASTE_SCRIPT_PATH "[FOLDER SCRIPT PATH]" and enter the location of the folder with the scripts between the "". Right-click and choose paste, to paste this location that was copied to the clipboard before. The command would then look for instance like this: SETX SETX EWASTE_SCRIPT_PATH "F:\Vincent\Documents\ewaste-master\scripts".

![Set script path](https://github.com/Statistics-Netherlands/ewaste/blob/master/images/set_script_path.png)

Now the DOS box can be closed. In case R was open, please close and re-open it. 

<br>

**R packages**

You need to install the following packages (sets of extra R functions) to be able to run the scripts. To do this open R and copy all of the next R commands - one at a time - into the R Console and press ENTER. If you are asked to select a CRAN mirror, select one of your choice and continue.

```r
install.packages("getopt")
install.packages("eurostat")
install.packages("plyr")
install.packages("dplyr")
install.packages("reshape2")
install.packages("readxl")
install.packages("data.table")
```

<br>
<br>
<br>

<a name="usage"/>
# Usage
The program can be run from the command line (Windows: cmd.exe, aka 'DOS box').

To run the calculations for all countries, go to the folder of the scrips. Do this by using the CD command. For instance 'cd /d F:\Vincent\Mijn documenten\ewaste-master\scripts'.
Then type the following text in the command editor:  
`Rscript main.R`

Alternatively the above two commands could be performed at once with the command 'Rscript %EWASTE_SCRIPT_PATH%\main.R'

To run the calculations just for one or more specific countries use:
`Rscript main.R --countries=nld;bel`. This will not perform calculations to correct outliers based on values from similar countries.


The above commands assume that the location where `Rscript.exe` lives is on the search path (the `PATH` environment variable).
If this is not the case (you will notice because you get a message saying "'Rscript' is not recognized as an internal or external command, operable program or batch file."), you will need to provide the full path to `Rscript.exe`, e.g. for R version 3.3.1 under MS Windows this would typically be:

* 32 bit computer:
C:\Program Files\R\R-3.3.1\bin\i386

* 64 bit computer:
C:\Program Files\R\R-3.3.1\bin\x64

For instance type '"C:\Program Files\R\R-3.3.1\bin\x64\Rscript.exe" main.R' (the path itself has to be between double-quotes because of the space in the name of the Program Files folder).
Then it looks like this:

![Start Rscript](https://github.com/Statistics-Netherlands/ewaste/blob/master/images/start_rscript.png)

<br>
In case you are comfortable with the R language you can of course also run the script from R. Executing the script main.R will give you the the calculations for all countries.
<br>
<br>
<br>

<a name="source-file-listing"/>
# Source file listing

* [00a_Prepare_Prodcom_data.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/00a_Prepare_Prodcom_data.R) - Create Prodcom data file
* [00b_Prepare_International_Trade_data.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/00b_Prepare_International_Trade_data.R) - Create international trade data file
* [00c_Prepare_Solar_Panel_data.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/00c_Prepare_Solar_Panel_data.R) - Create file with the weight and number of units of solar panels placed on the market
* [01_Prodcom_confidentials.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/01_Prodcom_confidentials.R) - Estimation of Prodcom confidential values
* [02_Apparent_Consumption_Method.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/02_Apparent_Consumption_Method.R) - Convert Prodcom data and International trade data to weight (kg) with the use of average weights data
* [03_POM_calculations.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03_POM_calculations.R) - remove outliers and make estimates for missing POM data
* [03a_remove_year_outliers.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03a_remove_year_outliers.R) - Outlier detection using Median Absolute Deviation (MAD)
* [03b_remove_stratum_outliers.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03b_remove_stratum_outliers.R) - Remove unreliable values (extremes) based on values of other years per Country and UNU_Key
* [03c_estimate_missings_using_years.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03c_estimate_missings_using_years.R) - Estimate missing values based on averages of surrounding years
* [03d_calculate_stratum_means.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03d_calculate_stratum_means.R) - Calculate stratum means for estimation of missing values based on stratum means
* [03e_estimate_missings_using_stratums.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03e_estimate_missings_using_stratums.R) - Estimate missing values based on stratum ratios
* [03f_smooth_years.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03f_smooth_years.R) - Eliminate big jumps from one year to the next
* [03g_CRT_monitors.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03g_CRT_monitors.R) - Change values in UNU_Key 0308 (CRT monitors)
* [03h_add_PV_panel_data.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03h_add_PV_panel_data.R) - Add Eurostat PV panel data (UNU_Key 0002) to use instead of the statistics
* [03i_extend_time_series](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03i_extend_time_series.R) - Extend time series back to 1980 and a few years into the future
* [03j_changes_compared_to_original_data.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/03j_changes_compared_to_original_data.R) - Look for effect of changes compared with the original source data
* [04_WEEE_calculations.R](https://github.com/Statistics-Netherlands/ewaste/blob/master/scripts/04_WEEE_calculations.R)] - Calculates the WEEE arising using the Weibull function

HISTORY.md
#### Release notes v1.2.

**00b_Prepare_International_Trade_data.R**

Switched to using Eurostat Bulkdownload Facility for International trade instead of the Easy Comext website.

**00c_Prepare_Solar_Panel_data.R**

Now using OData for collecting information about solar panels capacity.

**03i_extend_time_series.R**

Extrapolations are completely redone, now various new approaches for forecasting and also a new approach for extrapolations in the past.

**04_WEEE_calculations.R**

Calculation of number of pieces for WEEE is now based on the average weight in the year product was put on the market instead of on the average weight of the product in the year it became WEEE.

**05_Make_tblAnalysis.R and 05a_calculate_aggregates.R**

Calculation of number of pieces for POM, WEEE and STOCK is now also done for the aggregates of the 54 UNU_Keys.

**and more**

* Some links of Prodcom and CN to UNU_Keys have been updated
* Scale and Shape parameters for Weibul function in the WEEE calculations have been updated
* Average Weights have been updated
* Some manual corrections have been added

<br>
<br>
<br>

***
#### Release notes v1.1.

**00a_Prepare_Prodcom_data.R**

This script used the 'RODBC' package to read Excel files. In some cases the script could result in errors because of conflicts with 64 bits Excel drivers. Now a new method is being used using the 'readxl' package. This does not require Excel drivers anymore.


scripts/03_POM_calculations.R
# -----------------------------------------------------------------------------------------------------------
#
#   Name:           03POM_calculations.R
#
#   Description:    This script read the CSV file with the calculated POM data.
#                   Then it will remove outliers and make estimates for missing data.
#
#
#   Author:         V.M. van Straalen - Statistics Netherlands
#
# -----------------------------------------------------------------------------------------------------------


setwd(DATA_PATH)
options(stringsAsFactors=FALSE, warn=0, scipen=999, digits=4)

require(plyr)
require(reshape2)

# ----------------------------------------------------------
# UNU_countries: Read raw version of POM data
# ----------------------------------------------------------
UNU_countries <- read.csv("UNU_countries.csv",
                          colClasses = c("character", "character", "character",
                                         "numeric", "numeric"))

# Make copy so at the end of the calculations, we can see the impact of the alterations.
original_UNU_countries <- UNU_countries

# ----------------------------------------------------------
# Stratum: Read table with country stratums
# ----------------------------------------------------------
Stratum <- read.csv("tbl_Countries.csv",quote = "\"",
                    colClasses = c("character", "NULL", "numeric", "NULL"))

# Convert country codes to uppercase.
Stratum$Country <- toupper(Stratum$Country)


# ----------------------------------------------------------
# Population: Read population data
# ----------------------------------------------------------
Population <- read.csv("tbl_Data.csv", quote = "\"",
                       colClasses = c("character", "numeric", "numeric", "character"))

# Only keep data about number of inhabitants
Population <- Population[Population$Destination == "pop 17", ]

# Convert country codes to uppercase.
Population$Country <- toupper(Population$Country)

Population$Destination <- NULL
Population <- plyr::rename(Population,c("Value"="Inhabitants"))

# For many countries (mostly outside Europe) there is no old data about number of inhabitants.
# We estimate it by taking the first known value and apply the world population growth rate found at:
# http://www.worldometers.info/world-population/world-population-by-year/

tbl_world_population <- read.csv2("tbl_world_population.csv", quote = "\"",
                                  colClasses = c("character", "NULL", "numeric", "NULL",  "NULL", "NULL", "NULL") )

names(tbl_world_population)[2] <- "GrowthRate"

Population <- merge(Population, tbl_world_population,  all.x = TRUE)

sortorder <- order(Population$Country, -Population$Year)
Population <- Population[sortorder, ]

# Get data for previous row
Population$PrevYear <- c(NA, head(Population$Year, -1))
Population$PrevCountry <- c(NA, head(Population$Country, -1))

selection <- which( is.na(Population$Inhabitants) &
                      (Population$Year == Population$PrevYear -1) &
                      (Population$Country == Population$PrevCountry) )

# perform calculation
for (i in selection)
{Population[i, "Inhabitants"] <- Population[(i - 1), "Inhabitants"] /
  ( (Population[i-1, "GrowthRate"] + 100) /100)
}

rm(tbl_world_population)
Population$GrowthRate <- NULL
Population$PrevYear <- NULL
Population$PrevCountry <- NULL


# ----------------------------------------------------------
# PurPow: Read table with purchasing power per country
# ----------------------------------------------------------
PurPow <- read.csv("tbl_Data.csv", quote = "\"",
                   colClasses = c("character", "numeric", "numeric", "character"))

# Only keep data about the purchasing power parity
PurPow <- PurPow[PurPow$Destination == "ppp ph17", ]

# Convert country codes to uppercase.
PurPow$Country <- toupper(PurPow$Country)

PurPow$Destination <- NULL
PurPow <- plyr::rename(PurPow,c("Value"="PPP"))





# ----------------------------------------------------------
#   **** Perform the calculations ***
# ----------------------------------------------------------

# ----------------------------------------------------------
# UNU_countries - Preparations
# ----------------------------------------------------------

# Make records for each combination
myUNU_Keys <- as.data.frame(table(UNU_countries$UNU_Key))
names(myUNU_Keys)[1] <- "UNU_Key"
myUNU_Keys[2] <- NULL

myYears <- as.data.frame(table(UNU_countries$Year))
names(myYears)[1] <- "Year"
myYears[2] <- NULL

myCountries <- as.data.frame(table(UNU_countries$Country))
names(myCountries)[1] <- "Country"
myCountries[2] <- NULL

UNU_Key_Year <- merge(myUNU_Keys, myYears,  all = TRUE)
UNU_Key_Year_Country <- merge(UNU_Key_Year, myCountries,  all = TRUE)

UNU_countries <- merge(UNU_countries, UNU_Key_Year_Country,  by=c("UNU_Key", "Year", "Country"),  all = TRUE)


# Create flag variable and give value of 0 for original data and 1 for missing data.
UNU_countries$flag <- ifelse(is.na(UNU_countries$POM_kg) & is.na(UNU_countries$POM_pieces),1,0)


rm(myUNU_Keys, myYears, myCountries, UNU_Key_Year, UNU_Key_Year_Country)


# ----------------------------------------------------------
# UNU_countries: Perform some manual corrections
# ----------------------------------------------------------
# This file contains manual corrections and a few values that are not corrections, but including them here
# makes sure that that data will not be considered an outlier later in the process.
manualdata <- read.csv("manual_corrections.csv", quote = "\"",
                       colClasses = c("character", "character", "character", "numeric", "numeric",
                                      "NULL", "NULL", "NULL"))

# attach manual data to original data
UNU_countries <- merge(UNU_countries, manualdata,  by=c("UNU_Key", "Year", "Country"),  all.x = TRUE)

# overwrite original data with the manual data
selection <- !is.na(UNU_countries$POM_kg_new)
UNU_countries[selection, "POM_kg"] <- UNU_countries[selection, "POM_kg_new"]
UNU_countries[selection, "POM_pieces"] <- UNU_countries[selection, "POM_pieces_new"]
UNU_countries[selection, "flag"] <- 60

# Clean-up
UNU_countries$POM_kg_new <- NULL
UNU_countries$POM_pieces_new <-NULL
rm(manualdata)


# ----------------------------------------------------------
# UNU_countries: Add variables needed in the calculations
# ----------------------------------------------------------
# Attach number of inhabitants for each country
UNU_countries <- merge(UNU_countries, Population,  by=c("Country", "Year"),  all.x = TRUE)
UNU_countries <- merge(UNU_countries, Stratum,  by="Country",  all.x = TRUE)

rm(Stratum) # Population is needed later for the estimation of past and future values

# Calculate kg and pieces per inhabitant
UNU_countries$kpi = UNU_countries$POM_kg / UNU_countries$Inhabitants
UNU_countries$ppi = UNU_countries$POM_pieces / UNU_countries$Inhabitants

# For now the total POM_kg and POM_pieces are removed. They will be calculated again later in the script.
UNU_countries$POM_kg <- NULL
UNU_countries$POM_pieces <- NULL


# ----------------------------------------------------------
# UNU_countries: Some corrections on the original data based on knowledge of the market
# ----------------------------------------------------------
# CRT televisions have not been sold in recent years.
# For stratum 1 countries they are set to zero starting in 2007.
# For stratum 2 countries they are set to zero starting in 2008.
# For stratum 3 countries they are set to zero starting in 2009.

selection <- which (UNU_countries$UNU_Key == "0407" & UNU_countries$Stratum == 1 & UNU_countries$Year >= 2007)
UNU_countries[selection, "kpi"] <- 0
UNU_countries[selection, "ppi"] <- 0
UNU_countries[selection, "flag"] <- 52
selection <- which(UNU_countries$UNU_Key == "0407" & UNU_countries$Stratum == 2 & UNU_countries$Year >= 2008)
UNU_countries[selection, "kpi"] <- 0
UNU_countries[selection, "ppi"] <- 0
UNU_countries[selection, "flag"] <- 52
selection <- which(UNU_countries$UNU_Key == "0407" & UNU_countries$Stratum == 3 & UNU_countries$Year >= 2009)
UNU_countries[selection, "kpi"] <- 0
UNU_countries[selection, "ppi"] <- 0
UNU_countries[selection, "flag"] <- 52

# To make the transition more smoothly the previous year gets half the value of the year before that one.
# First sort dataframe rows by UNU_Key, Country and Year.
sortorder <- order(UNU_countries$UNU_Key, UNU_countries$Country, UNU_countries$Year)
UNU_countries <- UNU_countries[sortorder, ]

selection <- which(UNU_countries$UNU_Key == "0407" & UNU_countries$Stratum == 1 & UNU_countries$Year == 2006)
UNU_countries[selection, "kpi"] <- UNU_countries[(selection -1), "kpi"] / 2
UNU_countries[selection, "ppi"] <- UNU_countries[(selection -1), "ppi"] / 2
UNU_countries[selection, "flag"] <- 52

selection <- which(UNU_countries$UNU_Key == "0407" & UNU_countries$Stratum == 2 & UNU_countries$Year == 2007)
UNU_countries[selection, "kpi"] <- UNU_countries[(selection -1), "kpi"] / 2
UNU_countries[selection, "ppi"] <- UNU_countries[(selection -1), "ppi"] / 2
UNU_countries[selection, "flag"] <- 52

selection <- which(UNU_countries$UNU_Key == "0407" & UNU_countries$Stratum == 3 & UNU_countries$Year == 2008)
UNU_countries[selection, "kpi"] <- UNU_countries[(selection -1), "kpi"] / 2
UNU_countries[selection, "ppi"] <- UNU_countries[(selection -1), "ppi"] / 2
UNU_countries[selection, "flag"] <- 52

# In case flag value was set but calculation gives missing value, then reset the flag to 1.
selection <- which( UNU_countries$UNU_Key == "0407" & UNU_countries$flag == 52 & is.na(UNU_countries$kpi) )
# in case of empty vector don't do command
if (length(selection) > 0)
  {UNU_countries[selection, "flag"] <- 1}
  



# There are too many unrealistic data points in UNU_Key 0802. For now values of over 20 kg per inhabitant ar removed.
# Otherwise only the largest outliers get removed, but much smaller - however still very big - outliers remain.
# More research is needed to see if average weight per prodcom code could be improved.
selection <- which(UNU_countries$UNU_Key == "0802" & UNU_countries$kpi > 20 )
UNU_countries[selection, "kpi"] <- NA
UNU_countries[selection, "ppi"] <- NA
UNU_countries[selection, "flag"] <- 5 # This is same flag value as values removed because they are high outliers in 3a.



# ----------------------------------------------------------
# UNU_countries: Search for extremes in Year per Country and UNU_Key and remove them
# ----------------------------------------------------------

# threshold for how many times the mad a value can be away from the median to be considered an outlier
distance_mad <- 4

source(file.path(SCRIPT_PATH, "03a_remove_year_outliers.R"))


# ----------------------------------------------------------
# UNU_countries: Search for extremes in Country per Stratum, UNU_Key and Year and remove them
# ----------------------------------------------------------

if ( PROCESS_ALL_COUNTRIES ) {
   source(file.path(SCRIPT_PATH, "03b_remove_stratum_outliers.R"))
}

# ----------------------------------------------------------
# UNU_countries: Estimate missing values based on averages of surrounding years
# ----------------------------------------------------------

# Set inputfile for calculations. This way we can use same calculations later for missing stratum means.
mydf <- UNU_countries
mydf <- plyr::rename(mydf,c("UNU_Key"="catA", "Country"="catB"))

# Perform calculations and write in mydf.
source(file.path(SCRIPT_PATH, "03c_estimate_missings_using_years.R"))

# Set back to original names.
mydf <- plyr::rename(mydf,c("catA"="UNU_Key", "catB"="Country"))
UNU_countries <- mydf
rm(mydf)


# ----------------------------------------------------------
# UNU_countries: Calulate stratum means for estimation missing values based on stratum means
# ----------------------------------------------------------

# Calculate stratum means and write into strat_tot
if ( PROCESS_ALL_COUNTRIES ) {
  source(file.path(SCRIPT_PATH, "03d_calculate_stratum_means.R"))
}

# ----------------------------------------------------------
# strat_tot: Estimate missing stratum values based on surrounding years
# ----------------------------------------------------------

# Perform calculations and write in mydf
if ( PROCESS_ALL_COUNTRIES ) {
  # Set inputfile for calculations
  mydf <- strat_tot
  mydf <- plyr::rename(mydf,c("UNU_Key"="catA", "Stratum"="catB", "kpi_stratum"="kpi", "ppi_stratum"="ppi"))
  
  source(file.path(SCRIPT_PATH, "03c_estimate_missings_using_years.R"))
  
  # Copy data to UNU_countries
  strat_tot <- plyr::rename(mydf,c("catA"="UNU_Key", "catB"="Stratum", "kpi"="kpi_stratum", "ppi"="ppi_stratum"))
  # flag is not needed here
  strat_tot$flag <- NULL
  
  # Reshape stratum values to wide form.
  # For the following calculations we need for every record all stratum values available.
  # Estimation of missing values will be with the purchasing power in such a way that 
  # the estimated value will be between the 2 stratum means of 2 adjacent stratums.
  
  # KPI stratum
  strat_tot_wide_kpi <- dcast(strat_tot, UNU_Key + Year ~ Stratum, value.var = "kpi_stratum")
  names(strat_tot_wide_kpi)[3:ncol(strat_tot_wide_kpi)] <- 
    paste("kpi_s", names(strat_tot_wide_kpi)[3:ncol(strat_tot_wide_kpi)], sep = "")
  
  # PPI stratum
  strat_tot_wide_ppi <- dcast(strat_tot, UNU_Key + Year ~ Stratum, value.var = "ppi_stratum")
  names(strat_tot_wide_ppi)[3:ncol(strat_tot_wide_ppi)] <- 
    paste("ppi_s", names(strat_tot_wide_ppi)[3:ncol(strat_tot_wide_ppi)], sep = "")
  
  # merge together
  strat_tot_wide <- merge(strat_tot_wide_kpi, strat_tot_wide_ppi,
                          by=c("UNU_Key", "Year"), all.x = TRUE)
  
  # Merge stratum means with UNU_countries data
  UNU_countries <- merge(UNU_countries, strat_tot_wide,
                         by=c("UNU_Key", "Year"), all.x = TRUE)
  
  rm(mydf, strat_tot, strat_tot_wide_kpi, strat_tot_wide_ppi, strat_tot_wide)
}


# ----------------------------------------------------------
# UNU_countries: Estimate missing values based on stratum ratio's
# ----------------------------------------------------------

if ( PROCESS_ALL_COUNTRIES ) {
  source(file.path(SCRIPT_PATH, "03e_estimate_missings_using_stratums.R"))
}

# ----------------------------------------------------------
# UNU_countries: Eliminate big jumps from one year to the next
# ----------------------------------------------------------

# Make copy of dataset before elimination of big jumps.
# In case data is removed which cannot afterwards be estimated with surrounding years, we can put the
# original value back.
UNU_countries_before_smooth <- UNU_countries[ , c("UNU_Key", "Year", "Country", "kpi", "ppi")]
UNU_countries_before_smooth <- plyr::rename(UNU_countries_before_smooth,c("kpi"="kpi_before_smooth",
                                                                          "ppi"="ppi_before_smooth"))


source(file.path(SCRIPT_PATH, "03f_smooth_years.R"))


# ----------------------------------------------------------
# UNU_countries: For second time estimate missing values based on averages of surrounding years
# ----------------------------------------------------------
# This syntax will be run a second time because of the removal of big jumps from year to year in the previous step.

# Set inputfile for calculations. This way we can use same calculations later for missing stratum means.
mydf <- UNU_countries
mydf <- plyr::rename(mydf,c("UNU_Key"="catA", "Country"="catB"))

# Perform calculations and write in mydf.
source(file.path(SCRIPT_PATH, "03c_estimate_missings_using_years.R"))
# Set back to original names.
mydf <- plyr::rename(mydf,c("catA"="UNU_Key", "catB"="Country"))
UNU_countries <- mydf
rm(mydf)

# Now check if data has been removed in big jumps removal procedure, which was not estimated afterwards.
# If so, put the removed values back
UNU_countries <- merge(UNU_countries, UNU_countries_before_smooth,  all.x = TRUE)
selection <- is.na(UNU_countries$kpi) & !is.na(UNU_countries$kpi_before_smooth)
UNU_countries[selection, "kpi"] <- UNU_countries[selection, "kpi_before_smooth"]
UNU_countries[selection, "ppi"] <- UNU_countries[selection, "ppi_before_smooth"]
UNU_countries[selection, "flag"] <- UNU_countries[selection, "flag"] - 100

UNU_countries$kpi_before_smooth <- NULL
UNU_countries$ppi_before_smooth <- NULL
rm(UNU_countries_before_smooth)


# ----------------------------------------------------------
# UNU_countries: Change values in UNU_Key 0308 (CRT monitors)
# ----------------------------------------------------------

source(file.path(SCRIPT_PATH, "03g_CRT_monitors.R"))



# ----------------------------------------------------------
# UNU_countries: Extend time series
# ----------------------------------------------------------

source(file.path(SCRIPT_PATH, "03i_extend_time_series.R"))



# ----------------------------------------------------------
# UNU_countries: Add Eurostat PV panel data (UNU_Key 0002)
# ----------------------------------------------------------

source(file.path(SCRIPT_PATH, "03h_add_PV_panel_data.R"))



# ----------------------------------------------------------
# UNU_countries: Look for effect of changes compared with the original source data
# ----------------------------------------------------------

source(file.path(SCRIPT_PATH, "03j_changes_compared_to_original_data.R"))


# # ----------------------------------------------------------
# # UNU_countries: LED lamps didn't exist before 2007
# # ----------------------------------------------------------
# 
# selection <- which (UNU_countries$UNU_Key == "0505" & UNU_countries$Year <= 2006)
# UNU_countries[selection, "kpi"] <- 0
# UNU_countries[selection, "ppi"] <- 0
# UNU_countries[selection, "flag"] <- 52
# 
# # To make the transition more smoothly the previous year gets half the value of the year before that one.
# # First sort dataframe rows by UNU_Key, Country and Year.
# sortorder <- order(UNU_countries$UNU_Key, UNU_countries$Country, UNU_countries$Year)
# UNU_countries <- UNU_countries[sortorder, ]
# 
# selection <- which(UNU_countries$UNU_Key == "0505" & UNU_countries$Year == 2007)
# UNU_countries[selection, "kpi"] <- UNU_countries[(selection +1), "kpi"] / 2
# UNU_countries[selection, "ppi"] <- UNU_countries[(selection +1), "ppi"] / 2
# UNU_countries[selection, "flag"] <- 52
# 


# ----------------------------------------------------------
# UNU_countries: Last alterations to datafile
# ----------------------------------------------------------

# Calculate total weight (kg) and number of pieces out of the kilo's and units per inhabitant
# They have to be recalculated after the last corrections in the kpi and ppi.
UNU_countries$POM_kg = UNU_countries$kpi * UNU_countries$Inhabitants
UNU_countries$POM_pieces = UNU_countries$ppi * UNU_countries$Inhabitants

# Convert kg to tonnes.
UNU_countries$POM_t = UNU_countries$POM_kg / 1000
UNU_countries$POM_kg <- NULL


# ------------------------- PRODCOM FLAG -------------------------------
# Take the flag from the ProdCom estimations (script 01) and
# merge it with the flag of the POM calculations (script 03).

tbl_PCC <- read.csv("tbl_PCC.csv", quote = "\"", 
                    colClasses = c("character", "character", "character", "numeric", "numeric",
                                   "character", "numeric"))

htbl_PCC_Match_Key <- read.csv("htbl_PCC_Match_Key.csv", quote = "\"",
                               colClasses = c("character", "character", "character"))

tbl_PCC <- merge(tbl_PCC, htbl_PCC_Match_Key, by=c("PCC", "Year"), all.x = TRUE)

# remove missings
tbl_PCC[which (tbl_PCC$UNU_Key == ""), "UNU_Key"] <- NA
tbl_PCC <- tbl_PCC[!is.na(tbl_PCC$UNU_Key), ]
tbl_PCC <- tbl_PCC[!is.na(tbl_PCC$flag), ]

# Count number of flags per UNU_Key, Year, Country and flag-type.
prodcom_flags <- count(tbl_PCC[!is.na(tbl_PCC$UNU_Key), ][ , c("UNU_Key", "Year", "Country", "flag")])
names(prodcom_flags)[5] <- "n_flag"

# Change format
prodcom_flags <- dcast(prodcom_flags, UNU_Key + Year + Country ~ flag, value.var = "n_flag")

# Add data to UNU_countries
UNU_countries <- merge(UNU_countries, prodcom_flags,  by=c("UNU_Key", "Year", "Country"),  all.x = TRUE)

# If there has been estimations based on export/prodcom ratio of other years or other countries, 
# and flag value is zero, then alter it to flag it, to mark at least part of the value is based
# on a prodcom value estimation.
selection <- which( !is.na(UNU_countries$`1`) & UNU_countries$flag==0 )
UNU_countries[selection, "flag"] <- 41
selection <- which( !is.na(UNU_countries$`2`) & UNU_countries$flag==0 )
UNU_countries[selection, "flag"] <- 42
selection <- which( !is.na(UNU_countries$`3`) & UNU_countries$flag==0 )
UNU_countries[selection, "flag"] <- 43

UNU_countries$`0` <- NULL
UNU_countries$`1` <- NULL
UNU_countries$`2` <- NULL
UNU_countries$`3` <- NULL

rm(tbl_PCC)
rm(prodcom_flags)

# ------------------------- END PRODCOM FLAG -------------------------------



# Sort order for columns
sortorder_c <- c("Stratum", "Country", "UNU_Key", "Year", "POM_t", "POM_pieces",
                 "Inhabitants", "kpi",  "ppi", "flag")

# Sort dataframe rows by Country, UNU_Key and Year
sortorder <- order(UNU_countries$Country, UNU_countries$UNU_Key, UNU_countries$Year)
UNU_countries <- UNU_countries[sortorder, sortorder_c]

# Save data into tbl_POM.csv
write.csv(UNU_countries, file = "tbl_POM.csv", quote = TRUE, row.names = FALSE)
